eZ/Publish/Core/Repository/TrashService.php

```diff
@@ -86,16 +86,18 @@ public function __construct(
      */
     public function loadTrashItem($trashItemId)
     {
-        if ($this->repository->hasAccess('content', 'restore') !== true) {
-            throw new UnauthorizedException('content', 'restore');
-        }
-
         $spiTrashItem = $this->persistenceHandler->trashHandler()->loadTrashItem($trashItemId);
         $trash = $this->buildDomainTrashItemObject($spiTrashItem);
         if (!$this->repository->canUser('content', 'read', $trash->getContentInfo())) {
             throw new UnauthorizedException('content', 'read');
         }
 
+        // TODO: Need to check (integration tests + self + QA) how Role limitation will behave with this (as there is no location)
+        // we could pass trash as target (same goes for content/read above), but again would need to check how it will behave.
+        if (!$this->repository->canUser('content', 'restore', $trash->getContentInfo())) {
```

Comment:
Indeed. There is a risk this will fail with subtree limitation on Role. Trash item after all doesn't belong to subtree it was removed from...